components/mediators/cache/org.wso2.carbon.mediator.cache/src/main/java/org/wso2/carbon/mediator/cache/util/HttpCachingFilter.java

```diff
@@ -0,0 +1,189 @@
+        //Validate the TTL of the cached response.
+        if (maxAge > -1) {
+            long responseExpirationTime = cachedResponse.getResponseFetchedTime() + maxAge * 1000;
+            if (responseExpirationTime < System.currentTimeMillis()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * This method returns the ETag value.
+     *
+     * @param httpHeaders Http headers.
+     * @return ETag value.
+     */
+    private static String getETagValue(Map<String, Object> httpHeaders) {
+        if (httpHeaders.get(HttpHeaders.ETAG) != null) {
+            return String.valueOf(httpHeaders.get(HttpHeaders.ETAG));
+        }
+        return null;
+    }
+
+    /**
+     * This method sets the Age header.
+     *
+     * @param cachedResponse The cached response to be returned.
+     * @param msgCtx         The messageContext.
+     */
+    @SuppressWarnings("unchecked")
+    public static void setAgeHeader(CachableResponse cachedResponse,
+                                    org.apache.axis2.context.MessageContext msgCtx) {
+        Map excessHeaders = new MultiValueMap();
+        long responseCachedTime = cachedResponse.getResponseFetchedTime();
+        long age = Math.abs((responseCachedTime - System.currentTimeMillis()) / 1000);
+        excessHeaders.put(HttpHeaders.AGE, String.valueOf(age));
+
+        msgCtx.setProperty(NhttpConstants.EXCESS_TRANSPORT_HEADERS, excessHeaders);
+    }
+
+    /**
+     * Set the response fetched time in milliseconds.
+     *
+     * @param headers  Transport headers.
+     * @param response Response to be cached.
+     * @throws ParseException throws exception if exception happen while parsing the date.
+     */
+    public static void setResponseCachedTime(Map<String, String> headers, CachableResponse response) throws
+            ParseException {
+        long responseFetchedTime;
+        String fetchedTime;
+        if (headers != null && (fetchedTime = headers.get(HttpHeaders.DATE)) != null) {
+            SimpleDateFormat format = new SimpleDateFormat(CachingConstants.DATE_PATTERN);
+            Date d = format.parse(fetchedTime);
+            responseFetchedTime = d.getTime();
+        } else {
+            responseFetchedTime = System.currentTimeMillis();
+        }
+        response.setResponseFetchedTime(responseFetchedTime);
+    }
+
+    /**
+     * This method returns whether no-store header exists in the response.
+     *
+     * @param synCtx MessageContext with the transport headers.
+     * @return Whether no-store exists or not.
+     */
+    @SuppressWarnings("unchecked")
+    public static boolean isNoStore(MessageContext synCtx) {
+        Map<String, String> headers;
+        org.apache.axis2.context.MessageContext msgCtx = ((Axis2MessageContext) synCtx).getAxis2MessageContext();
+        headers = (Map<String, String>) msgCtx.getProperty("org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS");
```

Comment:
Shall we take a copy of TRANSPORT_HEADERS map here? This will take the reference to the TRANSPORT_HEADERS map and may cause ConcurrentModificationException in concurrent scenarios.